Feature/chadli

Abdelillah-CHADLI opened the pull request:
# YACK App - Chadli Updates Report
**Date:** December 2, 2025  
**Branch:** feature/chadli_  
**Repository:** generalBenmalek/YACK

---

## Summary

This document summarizes all updates made during the development session, integrating Isar database with the existing Hive + Firebase implementation and implementing multiple GitHub tasks.

---

## Architecture Overview

### Database Layer
| Component | Purpose |
|-----------|---------|
| **Isar** | Local reactive database for contracts, messages, media files, notifications |
| **Hive** | Encryption key storage (contractId → contractKey mapping) |
| **Firebase RTDB** | Online sync, invitation system, real-time updates |

### State Management
- **flutter_bloc** (Cubit pattern) for contract state management
- **TranslationHandler** for internationalization (en, ar, fr)

### Contract Status Flow
```
pending → accepted → completed
       → rejected
       → onDispute
```

---

## Completed Tasks

### 1. Isar-Firebase-Hive Integration
**File:** `lib/db/isar_adapter.dart`

Created a bridge layer that:
- Syncs contracts between Isar (local) and Firebase (remote)
- Manages encryption keys via Hive
- Handles offline-first data persistence

### 2. QR Contract Sharing System
**Files:**
- `lib/screens/shareContract.dart` - QR generation for contract sharing
- `lib/screens/scan_contract.dart` - QR scanning and contract acceptance

**Features:**
- Generate QR codes containing encrypted contract data
- Scan QR codes to receive shared contracts
- Silent notifications on contract acceptance
- Auto-navigation to home after scanning

### 3. Notifications Page (Rewritten)
**File:** `lib/screens/notifications.dart`

**Features:**
- Real-time updates using Isar watchers
- Multi-selection mode for bulk operations
- Mark as read / Delete selected
- Select all functionality
- Proper notification status display

### 4. Contract Agreement Page with Cubit
**Files:**
- `lib/screens/contract_agr/contract_agreement.dart` - Main contract view
- `lib/screens/contract_agr/contract_cubit.dart` - State management

**Contract Cubit Methods:**
```dart
loadContract()      // Load contract data
acceptContract()    // pending → accepted
rejectContract()    // pending → rejected  
completeContract()  // accepted → completed
disputeContract()   // any → onDispute
```

**Status Section Logic:**
| Current Status | Available Actions |
|----------------|-------------------|
| Pending | Accept, Reject, Dispute buttons |
| Accepted | Complete, Dispute buttons |
| Rejected | Status badge only |
| Completed | Status badge only |
| On Dispute | Status badge only |

### 5. No More Contracts Page
**File:** `lib/screens/contract_agr/nomore_contracts.dart`

**Updates:**
- Changed from network image to local asset: `assets/images/no_more_contracts.png`
- Added navigation to Subscription page on upgrade tap

### 6. Subscription Page (NEW)
**File:** `lib/screens/subscription.dart`

**Plans:**
| Plan | Price | Features |
|------|-------|----------|
| Basic | Free | 5 contracts/month, Basic support, Local storage |
| Pro | 2000 DA/month | 25 contracts/month, Priority support, Cloud backup, Templates |
| Unlimited | 10000 DA/month | Unlimited contracts, VIP support, Cloud sync, Advanced analytics, Team collaboration |

**Route:** `/subscription`

---

## Files Modified

### Core Files
| File | Changes |
|------|---------|
| `lib/main.dart` | Added subscription route, fixed duplicate import |
| `lib/db/isar_adapter.dart` | Created Isar-Firebase bridge |
| `lib/db/online.dart` | Removed unused cloudinary import |
| `lib/firebase_options.dart` | Removed unused firebase_database import |

### Screen Files
| File | Changes |
|------|---------|
| `lib/screens/contract_agr/contract_agreement.dart` | Fixed status section logic, added Complete button |
| `lib/screens/contract_agr/contract_cubit.dart` | Added reject/complete methods |
| `lib/screens/contract_agr/nomore_contracts.dart` | Local image, subscription navigation |
| `lib/screens/subscription.dart` | **NEW** - Subscription plans page |
| `lib/screens/notifications.dart` | Isar watchers, multi-selection |
| `lib/screens/scan_contract.dart` | Removed unused import |
| `lib/screens/shareContract.dart` | QR sharing implementation |
| `lib/screens/root.dart` | Removed unused import |

### Translation Files
| File | Keys Added |
|------|------------|
| `lib/utils/languages/en.dart` | new keys |
| `lib/utils/languages/fr.dart` | new keys |
| `lib/utils/languages/ar.dart` | new keys |

**New Translation Keys:**
- Contract status: `status_disputed`, `status_pending`, `status_rejected`
- Actions: `complete`, `complete_contract`, `complete_contract_warning`
- Subscription: `subscription`, `choose_your_plan`, `unlock_premium_features`
- Plans: `plan_basic`, `plan_pro`, `plan_unlimited`
- Features: `feature_5_contracts`, `feature_25_contracts`, `feature_unlimited_contracts`, etc.
- UI: `popular`, `current_plan`, `subscribe_now`, `cancel_anytime`, `subscription_success`, `subscribed_to`

---

## Asset Configuration

**pubspec.yaml:**
```yaml
flutter:
  assets:
    - assets/images/
```

**Required Image:**
- `assets/images/no_more_contracts.png`

---

## Routes

| Route | Screen | Description |
|-------|--------|-------------|
| `/welcome` | OnboardingScreen | App introduction |
| `/signup` | SignUpScreen | User registration |
| `/login` | LoginScreen | User authentication |
| `/confirm` | ConfirmAccount | Email verification |
| `/forgot-password` | ForgetPassword | Password reset |
| `/contract/scan_contract` | ScanContractScreen | QR code scanning |
| `/contract/sign_contract` | SignContractScreen | Contract signing |
| `/contract/create_contract` | CreateContractScreen | Create new contract |
| `/contract/view` | ContractAgreement | View contract (requires contractId arg) |
| `/settings` | SettingsScreen | App settings |
| `/upgrade` | NoMoreContractsAvailable | Contract limit reached |
| `/subscription` | SubscriptionScreen | **NEW** - Choose subscription plan |
| `/home` | BottomNavBar | Main app home |

---

## System State Verification

### Expected Working Features:
1. User authentication (login/signup/password reset)
2. Contract creation and management
3. QR code contract sharing
4. Contract scanning and acceptance
5. Real-time notifications with Isar
6. Contract status management (accept/reject/complete/dispute)
7. Multi-language support (EN/FR/AR)
8. Subscription page with plan selection
9. No More Contracts page with local image
10. Offline-first data persistence

---

-> Bypassed authentication for testing, now just merge the authentication with this version of this app and everything is working (real-time database, offline database) (fully tested)

-> Some messages need translation (didn't add them all)

**Report Created:** December 2, 2025  (7:32 AM)  
**Status:** All requested features implemented and working


-- Update

- Update compileSdk and targetSdk to 36
- Simplify build.gradle.kts (remove afterEvaluate blocks causing evaluation errors)
- Add gradle.properties settings for Android build features
- Upgrade isar packages to ^3.1.0+1
- Regenerate Isar model files for new isar version

Note: isar_flutter_libs requires manual patching in pub cache:
- Add namespace 'dev.isar.isar_flutter_libs' to build.gradle
- Update compileSdkVersion to 36"

yacinearezki committed:
back end done

yacinearezki committed:
fully done backend

Abdelillah-CHADLI committed:
feat: major bug fixes and improvements for contract chat system

Contract Completion: fix auto-confirm bugs, add persistent sync, proper decline handling

Chat: fix message duplication with firebaseKey deduplication

Contracts: immediate Firebase registration on acceptance

Media: sync support for Cloudinary URLs

Real-time: dispute, close request, and completion notifications

Abdelillah-CHADLI committed:
fix(android): update build config for AGP 8.x compatibility

generalBenmalek commented:
I tested it and it is working
 I will merge it later with the auth system to make the complete software we need for this submission

generalBenmalek reviewed (dismissed):
excellent job

generalBenmalek committed:
Merge branch 'develop' into feature/chadli_

generalBenmalek committed:
Add unverified user state and update contract view routing

Introduces an UnverifiedUser state to handle users who are authenticated but have not verified their email. Updates AuthService.isAuthenticated to return null for unverified users and adjusts AuthCubit logic accordingly. Modifies contract view route to accept a contractId argument. Cleans up duplicate and unused translation keys in English and French language files.

generalBenmalek committed:
Redirect unverified users to confirmation page

Adds a check for the UnverifiedUser state in AppWrapper and redirects such users to the '/confirm' route using Navigator.pushReplacementNamed.
